applications/FluidDynamicsApplication/custom_elements/compressible_navier_stokes_explicit.h

```diff
@@ -508,6 +513,550 @@ class CompressibleNavierStokesExplicit : public Element
+
+        for (unsigned int i_gauss = 0; i_gauss < r_integration_points.size(); ++i_gauss) {
+            rOutput[i_gauss] = temp_grad;
+        }
+    } else if (rVariable == VELOCITY_ROTATIONAL) {
+        const array_1d<double,3> rot_v = CalculateMidPointVelocityRotational();
+        for (unsigned int i_gauss = 0; i_gauss < r_integration_points.size(); ++i_gauss) {
+            rOutput[i_gauss] = rot_v;
+        }
+    } else {
+        KRATOS_ERROR << "Variable not implemented." << std::endl;
+    }
+}
+
+/**
+ *  This namespace is used to implement the compile-time choice of:
+ *  1. Shape function calculator
+ *  2. Element size calculator
+ *
+ *  [1] This is necessary because simplex geometries' shape functions are known
+ *  at compile-time whereas non-simplex's ones are not (because they depend on
+ *  the actual shape of the element).
+ *
+ *  [2] This is useful because GradientsElementSize is only implemented for
+ *  simplex types.
+ *
+ *  This problem is solved using SFINAE
+ */
+namespace CompressibleNavierStokesExplicitInternal
+{
+    template<unsigned int TDim, unsigned int TNumNodes>
+    using ElementDataStruct = typename CompressibleNavierStokesExplicit<TDim, TNumNodes>::ElementDataStruct;
+
+
+    constexpr bool IsSimplex(const unsigned int dimensions, const unsigned int nnodes)
+    {
+        return dimensions == nnodes-1;
+    }
+
+    // If is simplex, it becomes: using EnableIfSimplex = T
+    // Otherwise it causes a Substitution Failure
+    template<unsigned int TDim, unsigned int TNumNodes, typename T>
+    using EnableIfSimplex = typename std::enable_if<IsSimplex(TDim, TNumNodes), T>::type;
+
+    // The oposite of EnableIfSimplex
+    template<unsigned int TDim, unsigned int TNumNodes, typename T>
+    using EnableIfNotSimplex = typename std::enable_if<!IsSimplex(TDim, TNumNodes), T>::type;
+
+    // Specialization for simplex geometries
+    template<unsigned int TDim, unsigned int TNumNodes>
+    EnableIfSimplex<TDim, TNumNodes, void> ComputeGeometryData(
+        const Geometry<Node<3>> & rGeometry,
+        ElementDataStruct<TDim, TNumNodes>& rData)
+    {
+        GeometryUtils::CalculateGeometryData(rGeometry, rData.DN_DX, rData.N, rData.volume);
+        rData.h = ElementSizeCalculator<TDim, TNumNodes>::GradientsElementSize(rData.DN_DX);
+    }
+
+    /**
+     * Specialization for geometries with non-uniform jacobian (non-simplex)
+     * Shape functions cannot be obtained here. They will need to be calculated
+     * during integration at each gauss point.
+     */
+    template<unsigned int TDim, unsigned int TNumNodes>
+    EnableIfNotSimplex<TDim, TNumNodes, void> ComputeGeometryData(
+        const Geometry<Node<3>> & rGeometry,
+        ElementDataStruct<TDim, TNumNodes>& rData)
+    {
+        switch (TDim) {
+            case 2: rData.volume = rGeometry.Area();   break;
+            case 3: rData.volume = rGeometry.Volume(); break;
+        }
+
```

Comment:
```suggestion
        rData.volume = rGeometry.DomainSize();
```
`DomainSize()` returns volume for 3D geometries and area for 2D ones.